src/lib/threadpool.c

```diff
@@ -0,0 +1,549 @@
+ *    !empty(u) && +-------  ^ |
+ * type(head(o)) != BAR      | |
+ *                           | | type(head(o)) == BAR
+ *            in_flight == 0 | |
+ *                           | V
+ *                         BARRIER --------+ signal()
+ *                           ^ |   <-------+
+ *                           | |
+ *                  empty(u) | | !empty(u)
+ *                           | V
+ *                      DRAINING_UNORD
+ */
+
+enum planner_states {
+	PS_NOTHING,
+	PS_DRAINING,
+	PS_BARRIER,
+	PS_DRAINING_UNORD,
+	PS_EXITED,
+};
+
+static const struct sm_conf planner_states[SM_STATES_MAX] = {
+       [PS_NOTHING] = {
+               .flags   = SM_INITIAL,
+               .name    = "nothing",
+               .allowed = BITS(PS_DRAINING)
+			| BITS(PS_EXITED),
+       },
+       [PS_DRAINING] = {
+               .name    = "draining",
+               .allowed = BITS(PS_DRAINING)
+			| BITS(PS_NOTHING)
+			| BITS(PS_BARRIER),
+       },
+       [PS_BARRIER] = {
+               .name    = "barrier",
+               .allowed = BITS(PS_DRAINING_UNORD)
+			| BITS(PS_DRAINING)
+			| BITS(PS_BARRIER),
+       },
+       [PS_DRAINING_UNORD] = {
+               .name    = "draining-unord",
+               .allowed = BITS(PS_BARRIER)
+       },
+       [PS_EXITED] = {
+               .flags   = SM_FINAL,
+               .name    = "exited",
+               .allowed = 0,
+       },
+};
+
+enum {
+	POOL_THREADPOOL_SIZE = 4,
+	MAX_THREADPOOL_SIZE  = 1024,
+	POOL_LOOP_MAGIC      = 0x00ba5e1e55ba5500, /* baseless bass */
+};
+
+typedef struct pool_thread pool_thread_t;
+typedef struct pool_impl pool_impl_t;
+
+struct targs {
+	pool_impl_t *pi;
+	uv_sem_t    *sem;
+	uint32_t     idx;       /* Thread's index */
+};
+
+/* Worker thread of the pool */
+struct pool_thread {
+	queue        inq;       /* Thread's input queue */
+	uv_cond_t    cond;      /* Signalled when work item appears in @inq */
+	uv_thread_t  thread;    /* Pool's worker thread */
+	struct targs arg;
+};
+
+struct pool_impl {
+	uv_mutex_t     mutex;           /* Input queue, planner_sm,
+                                           worker and planner threads lock */
+	uint32_t       nthreads;
+	pool_thread_t *threads;
+
+	queue          outq;            /* Output queue used by libuv part */
+	uv_mutex_t     outq_mutex;      /* Output queue lock */
+	uv_async_t     outq_async;      /* Signalled when output queue is not
+                                           empty and libuv loop has to process
+                                           items from it */
+	uint64_t       active_ws;       /* Number of all work items in flight */
+
+	queue          ordered;         /* Queue of WT_ORD{N} items */
+	queue          unordered;       /* Queue of WT_UNORD items */
+	struct sm      planner_sm;      /* State machine of the scheduler */
+	uv_cond_t      planner_cond;
+	uv_thread_t    planner_thread;  /* Scheduler's thread */
+
+	uv_key_t       thread_key;      /* TLS: stores thread id */
```

Comment:
Is there a use-case for storing the thread ID other than to enable testing? Not a big deal either way but I'm curious